src/codegen.cpp

```diff
@@ -3220,7 +3226,25 @@ static jl_cgval_t emit_expr(jl_value_t *expr, jl_codectx_t *ctx)
                 return mark_julia_const(c);
             }
         }
-        return emit_call(ex, ctx);
+        jl_cgval_t res = emit_call(ex, ctx);
+        // some intrinsics (e.g. typeassert) can return a wider type
+        // than what's actually possible
+        jl_value_t *expr_t = expr_type((jl_value_t*)ex, ctx);
+        if (res.typ == (jl_value_t*)jl_any_type || expr_t == jl_bottom_type) {
+            res.typ = expr_t;
```

Comment:
reassigning `res.typ` isn't valid and is likely to lead to later assertion errors (there's `remark_julia_type`, which is a bit safer by at least keeping the cgval object in a valid state, but that's usually wrong to call still)